File: js/test_money.js
const assert = require('assert');
const Money = require('./money');
const Portfolio = require('./portfolio');
const Bank = require('./bank');

class MoneyTest {
    setUp() {
        this.bank = new Bank();
        this.bank.addExchangeRate("EUR", "USD", 1.2);
        this.bank.addExchangeRate("USD", "KRW", 1100);
    }

    testMultiplication() {
        let tenEuros = new Money(10, 'EUR');
        let twentyEuros = new Money(20, 'EUR');
        assert.deepStrictEqual(tenEuros.times(2), twentyEuros);
    }

    testDivision() {
        let originalMoney = new Money(4002, 'KWR');
        let moneyAfterDivision = originalMoney.divide(4);
        let expectmoneyAfterDivison = new Money(1000.5, 'KWR')
        assert.deepStrictEqual(moneyAfterDivision, expectmoneyAfterDivison);
    }

    testAddition(){
        let fiveDollars = new Money(5, 'USD');
        let tenDollars = new Money(10, 'USD');
        let fifteenDollars = new Money(15, 'USD');
        let portfolio = new Portfolio();
        portfolio.add(fiveDollars, tenDollars);
        assert.deepStrictEqual(portfolio.evaluate(this.bank, 'USD'), fifteenDollars);
    }

    testAdditionOfDollarsAndEuros() {
        let fiveDollars = new Money(5, 'USD');
        let tenEuros = new Money(10, 'EUR');
        let portfolio = new Portfolio();
        portfolio.add(fiveDollars, tenEuros);
        let expectedValue = new Money(17, 'USD');
        assert.deepStrictEqual(portfolio.evaluate(this.bank, 'USD'), expectedValue);
    }

    testAdditionOfDollarsAndWons() {
        let oneDollar = new Money(1, 'USD');
        let elevenHundredWon = new Money(1100, 'KRW');
        let portfolio = new Portfolio();
        portfolio.add(oneDollar, elevenHundredWon);
        let expectedValue = new Money(2200, 'KRW');
        assert.deepStrictEqual(portfolio.evaluate(this.bank, 'KRW'), expectedValue);
    }

    testAdditionWithMultipleMissingExchangeRates() {
        let oneDollar = new Money(1, 'USD');
        let oneEuro = new Money(1, 'EUR');
        let oneWon = new Money(1, 'KRW');
        let portfolio = new Portfolio();
        portfolio.add(oneDollar, oneEuro, oneWon);
        let expectedError = new Error(
            "Missing exchange rate(s):[USD->Kalganid,EUR->Kalganid,KRW->Kalganid]"
        )
        assert.throws(() => {portfolio.evaluate(this.bank, 'Kalganid')}, expectedError);
    }

    testConversionWithDifferentRatesBetweenTwoCurrencies(){
        let bank = new Bank();
        bank.addExchangeRate('EUR', 'USD', 1.2);
        let tenEuros = new Money(10, 'EUR');
        assert.deepStrictEqual(
            bank.convert(tenEuros, 'USD'), new Money(12, 'USD')
        );

        this.bank.addExchangeRate("EUR", "USD", 1.3); 2
        assert.deepStrictEqual(this.bank.convert(tenEuros, "USD"),
        new Money(13, "USD"));
    }

    testConversionWithMissingExchangeRates(){
        let bank = new Bank();
        let tenEuros = new Money(10, 'EUR');
        let expectedError = new Error('EUR->Kalganid');
        assert.throws(function() {bank.convert(tenEuros, 'Kalganid')}, expectedError);
    }

    testWhatIsConversionRateFromEURToUSD(){
        let tenEuros = new Money(10, 'EUR');
        assert.deepStrictEqual(this.bank.convert(tenEuros, 'USD'), 
        new Money(12, 'USD'))
    }

    getAllTestMethods() {
        let moneyPrototype = MoneyTest.prototype;
        let allprops = Object.getOwnPropertyNames(moneyPrototype);
        let testMethods = allprops.filter(p => {
            return typeof moneyPrototype[p] === 'function' && p.startsWith('test');
        })
        return testMethods;
    }

    runAllTests() {
        let testMethods = this.getAllTestMethods();
        testMethods.forEach(m => {
            console.log("Running: %s()", m);
            let method = Reflect.get(this, m);
            try{
                this.setUp();
                Reflect.apply(method, this, []);
            } catch (e){
                if (e instanceof assert.AssertionError) {
                    console.log(e);
                } else {
                    throw e;
                }
            }
        })
    }
}

new MoneyTest().runAllTests();

File: py/test_money.py
import unittest
from money import Money
from portfolio import Portfolio
from bank import Bank

class TestMoney(unittest.TestCase):
    def setUp(self) -> None:
        self.bank = Bank()
        self.bank.addExchangeRate("EUR", "USD", 1.2)
        self.bank.addExchangeRate("USD", "KRW", 1100)

    def testMultiplication(self):
        tenEuros = Money(10, 'EUR')
        twentyEuros = Money(20, 'EUR')
        self.assertEqual(twentyEuros, tenEuros.times(2))

    def testDivsion(self):
        originalMoney = Money(4002, 'KWR')
        expectedMoneyAfterDivision = Money(1000.5, 'KWR')
        self.assertEqual(expectedMoneyAfterDivision, originalMoney.divide(4))

    def testAddition(self):
        fiveDollars = Money(5, 'USD')
        tenDollars = Money(10, 'USD')
        fifteenDollars = Money(15, 'USD')
        portfolio = Portfolio()
        portfolio.add(fiveDollars, tenDollars)
        self.assertEqual(fifteenDollars, portfolio.evaluate(self.bank, 'USD'))

    def testAdditonOfDollarsAndEuros(self):
        fiveDollars = Money(5, 'USD')
        tenEuros = Money(10, 'EUR')
        portfolio = Portfolio()
        portfolio.add(fiveDollars, tenEuros)
        expectedValue = Money(17, 'USD')
        self.assertEqual(expectedValue, portfolio.evaluate(self.bank, 'USD'))

    def testAdditionOfDollarsAndWon(self):
        oneDollar = Money(1, 'USD')
        elevenHundredWons = Money(1100, 'KRW')
        portfolio = Portfolio()
        portfolio.add(oneDollar, elevenHundredWons)
        expectedValue = Money(2200, 'KRW')
        self.assertEqual(expectedValue, portfolio.evaluate(self.bank, 'KRW'))

    def testAdditionWithMultipleMissingExchangeRates(self):
        oneDollar = Money(1, 'USD')
        oneEuro = Money(1, 'EUR')
        oneWon = Money(1, 'KRW')
        portfolio = Portfolio()
        portfolio.add(oneDollar, oneEuro, oneWon)
        with self.assertRaisesRegex(
            Exception,
            "Missing exchange rate\(s\):\[USD\->Kalganid,EUR->Kalganid,KRW->Kalganid]",
        ):
            portfolio.evaluate(self.bank, "Kalganid")

    def testConversion(self):
        bank = Bank()
        bank.addExchangeRate('EUR', 'USD', 1.2)
        tenEuros = Money(10, 'EUR')
        self.assertEqual(bank.convert(tenEuros, 'USD'), Money(12, 'USD'))

    def testConversionWithMissingExchangeRates(self):
        bank = Bank()
        tenEuros = Money(10, 'EUR')
        with self.assertRaisesRegex(Exception, "EUR->Kalganid"):
            bank.convert(tenEuros, 'Kalganid')

if __name__ == '__main__':
    unittest.main()